docs/config.rst
Configuration
=============

The defaults described are the defaults in code and not necessarily what is in the default orlo.ini.

[main]
``````

:debug_mode: `true` or `false`. Default `false`. Enables Flask's debug mode.
:time_format: A `strftime <https://docs.python.org/2/library/time.html#time.strftime>`_
    string. Default `%Y-%m-%dT%H:%M:%SZ`.
:time_zone: Local time zone, as understood by pytz. Internally,
    all timestamps are stored in UTC. The timestamp is interpreted by Arrow when
    timestamps are given by the user (e.g. on recording a release), this
    setting merely reflects what time zone is given back to GET requests.
    Default `UTC`.
:base_url: The external url which points to your web app. Required for
    callbacks. Default `http://localhost:8080`.

[security]
``````````

:enabled: `true` or `false`. Enables security.
:passwd_file: Path to a `htpasswd <https://httpd.apache.org/docs/2.2/programs/htpasswd.html>`_
    file to use for authentication.
:secret_key: A secret key to use for token encryption. Default `change_me`.
    If security is enabled and this is still set to `change_me`, Orlo will
    refuse to start.
:token_ttl: The length of times that tokens should live for in seconds.
    Tokens automatically expire after this. Default `3600`.
:ldap_server: Ldap server to use for ldap requests. Default `localhost
    .localdomain`
:ldap_port: Ldap port to use for ldap requests. Default `389`.
:user_base_dn: Ldap dn in which to search for users. Default `ou=people,
    ou=example,ou=test`


[db]
````

:uri: Database uri for SQLAlchemy. See `Flask-SQLAlchemy <http://flask-sqlalchemy.pocoo.org/2.1/config/?highlight=sqlalchemy_database_uri>`_
    docs for details. Default `postgres://orlo:password@localhost:5432/orlo`
:echo_queries: Whether or not to echo sql queries to log. Default `false`.
:pool_size: Pool size for database connections. Default `50`. 
    This default errs on the high side and could probably be reduced for most installations.

[flask]
```````
:strict_slashes: `true` or `false`. Default `false`. By default, Werkzeug
    (what Flask uses underneath), will automatically "handle" trailing slashes,
    with the result that /foo/ and /foo are the same url. This disables that
    behaviour. It is recommended that you leave this set to false. See the
    `Werkzeug documentation <http://werkzeug.pocoo.org/docs/0.11/routing/#maps-rules-and-adapters>`_
    for more information.
:propagate_exceptions: `true` or `false`. Default `true`. Sets
    'PROPAGATE_EXCEPTIONS` in Flask. See
    `Flask documentation <http://flask.pocoo.org/docs/0.10/config/#builtin-configuration-values>`_
    for details.
:debug: Enable Flask debug mode. Default `false`. This is a security risk, enable with care.


[gunicorn]
``````````
:workers: Number of gunicorn workers to start (for handling requests).
:bind: Address:port to bind to. Default `127.0.0.1:8080`.

[logging]
`````````

:level: Logging level, valid values `debug`, `info`, `warning`, `error`.
    Default `info`.
:directory: Log directory to store logs in, if logging is enabled.
:format: Output format of logs. This should be a string which is accepted by Python's logging.Formatter.
    See `Formatter Objects <https://docs.python.org/3.6/library/logging.html#formatter-objects>`_
    and `LogRecord Attributes <https://docs.python.org/3.6/library/logging.html#logrecord-attributes>`_ for more information.
    Default `%(asctime)s [%(name)s] %(levelname)s %(module)s:%(funcName)s:%(lineno)d - %(message)s`


debian/changelog
orlo (0.4.1) jessie; urgency=medium

  Build changes only

  [ Alex Forbes ]
  * postgressql-server-dev-all needed for build
  * Add /usr/local/bin to build environment path
  * Preinstall no-manylinux package to avoid manylinux wheels
  * Include orloclient 0.4.0
  * Mock psycopg2, pyldap and gunicorn in docs build

 -- Alex Forbes <github@al4.co.nz>  Tue, 04 Apr 2017 15:39:06 +0000

orlo (0.4.0) jessie; urgency=medium

  [ Alex Forbes ]
  * Fix exec path
  * Restrict gbp dch to debian dir, add debuild artifacts to gitignore
  * Set universal wheel in setup.cfg
  * Add orloclient as dependency
  * Add /internal/version endpoint
  * Move routes to subdirectory
  * Refactor orlo command line, embed gunicorn app, add db migrations
  * Commit initial database version
  * WIP
  * Fix migration script
  * Fix up log handling
  * Alembic migration WIP
  * Use extras hack for tests_require
  * Update vagrantfile
  * Add libpq-dev to vagrantfile, required for psycopg2
  * Upgrade pip/setuptools in virtualenv
  * Depend on orloclient 0.2.0 for now, until 0.3.0 is released
  * Add release_id to package dictionaries
  * Remove /deploy functionality
  * Remove deployer-related files
  * Configure postgres in vagrant image
  * Stamp pre-existing databases with the initial revision
  * Remove checkfirst from create_table statements
  * Comment out postgres config for now
  * Suggested changes, courtesy of Carlo Bongiovanni
  * Remove obsolete and empty test file
  * Allow overriding of log directory by environment
  * Upgrade postgres to 9.4 in travis
  * Remove unneeded requirements.txt files
  * Use trusty travis image rather than precise
  * Increase tox verbosity
  * Fix packages entry in setup.py
  * Bump version to 0.4.0

 -- Alex Forbes <github@al4.co.nz>  Thu, 24 Nov 2016 10:09:28 +0000

orlo (0.3.1-2) jessie; urgency=medium

  [ Alex Forbes ]
  * Fix path to gunicorn in systemd unit

  [ vagrant ]

 -- Alex Forbes <github@al4.co.nz>  Wed, 23 Nov 2016 18:02:42 +0000

orlo (0.3.1-1) jessie; urgency=medium

  [ Alex Forbes ]
  * Fix test
  * Rename __init__ to test_base and update imports
  * Create multiple vagrant machines for testing
  * Add jessie and xenial vagrant boxes for build/testing
  * Run python setup.py clean in make clean
  * Change deb install location to /opt/venvs
  * Add initial tox.ini
  * Default internal config should be sqlite
  * Configure tests to use postgres in vagrant
  * Remove trusty vm, fix IP
  * Use main repos, upgrade dh-virtualenv, fix missing \
  * Add db clients and restart postgres
  * Small doc fix
  * Remove dependency on passwd file by mocking the verification function
  * Use tox to run tests in travis
  * mock has moved under py3
  * Set test database to localhost if running in Travis
  * Set posargs for tox and maxfail for pytest
  * Pass travis env car through to tox
  * Rename DeployTest as LiveDbTest, move to test_base
  * Add stress test
  * Add liveserver config to app, refactor db test with classmethods
  * Close the session in our json streamer to avoid leaking connections
  * Test refactor
  * Fix test failure caused by previous tests not clearing db connection
  * Bump version to 0.3.1
  * Fix locale in vagrantfile (for me anyway 😁), use tox for testing

  [ vagrant ]

 -- Alex Forbes <github@al4.co.nz>  Wed, 23 Nov 2016 17:39:48 +0000

orlo (0.3.1) trusty; urgency=medium

  * Add support for filtering by reference

 -- Alex Forbes <github@al4.co.nz>  Tue, 05 Jul 2016 17:24:17 +0000

orlo (0.3.0) trusty; urgency=medium

  * Make vagrant owner of /vagrant for builds
  * Reorganise tests
  * Add basic /packages endpoint
  * Bump version
  * Fix /packages test

 -- Alex Forbes <github@al4.co.nz>  Tue, 05 Jul 2016 15:19:56 +0000

orlo (0.2.0) trusty; urgency=medium

  [ Alex Forbes ]
  * Add dependant python virtualenv package to Vagrantfile
  * Add outline for Deploy process
  * Use release object instead of re-declaring, add live server test

  [ Ivan Coppa ]
  * Support for Release metadata attributes, Deploy endpoint
  * disabling debug

  [ Alex Forbes ]
  * Add integrated dummy deployer test
  * Use relative path for deployer.rb

  [ Marcel Kuiper ]
  * Added POC for Token based AUthentication

  [ Ivan Coppa ]
  * enable deploy api call

  [ Marcel Kuiper ]
  * processed flake8 recommendations on user_auth.py

  [ ivancoppa ]
  * create base data

  [ Alex Forbes ]
  * Add last minute changes to create script
  * Add last minute hacks to make everything work
  * Add orlo_release to deployer env
  * Create configuration for auth +
  * Make URL a parameter for auth tests
  * Only apply login_required when security is enabled
  * Rename test classes to start with "Test"
  * Move auth config into setup/teardown
  * Add root page, test fixes
  * Skip deploy test for now
  * Move route_api to route_releases, rename metadatas => metadata
  * Make token lifetime configurable, update orlo.ini with new options
  * Bump version of Flask-HTTPAuth
  * Add status code to OrloAuthError and add OrloConfigError
  * Remove login_handler in favour or separate TokenAuth module
  * Seperate token auth from http basic auth
  * Only require auth when configured
  * Implement deployer.py test deployer script
  * Bump orloclient version, use py.test for travis
  * Add pytz to requirements.txt
  * Add requirements_testing.txt
  * Move pytest to requirements_testing
  * Use /bin/true in deploy shell test

  [ Dustin Nguyen ]
  * Added ldap auth
  * Added ldap mock test

  [ Alex Forbes ]
  * Misc changes to deploy
  * Tidy up example deployer
  * Bump to pre-release 0.2
  * Tidy comments in example deployer
  * Use absolute path for version file
  * Use orlo.cli as entry point instead of custom script
  * Update Vagrantfile
  * Fix silly bug, read config from file after set
  * Move debian packaging files in to debian dir
  * Explicitly set synced_folder to virtualbox provider
  * [lintian] fix description
  * Add /version url and test

orlo/config.py
from __future__ import print_function
import os
from six.moves.configparser import RawConfigParser

__author__ = 'alforbes'


# Defaults that can be overridden by environment variables
defaults = {
    'ORLO_CONFIG': '/etc/orlo/orlo.ini',
    'ORLO_LOGDIR': '/var/log/orlo',
}

for var, default in defaults.items():
    try:
        defaults[var] = os.environ[var]
    except KeyError:
        pass

# Horrible hack, but I am tired
if os.environ.get('READTHEDOCS', None) == 'True':
    defaults['ORLO_LOGDIR'] = 'disabled'

config = RawConfigParser()

config.add_section('main')
config.set('main', 'time_format', '%Y-%m-%dT%H:%M:%SZ')
config.set('main', 'time_zone', 'UTC')
config.set('main', 'base_url', 'http://localhost:8080')

config.add_section('gunicorn')
config.set('gunicorn', 'workers', '2')
config.set('gunicorn', 'bind', '127.0.0.1:8080')

config.add_section('security')
config.set('security', 'enabled', 'false')
config.set('security', 'passwd_file', 'none')
config.set('security', 'secret_key', 'change_me')
# NOTE: orlo.__init__ checks that secret_key is not "change_me" when security
# is enabled. Do not change the default here without updating __init__ as well.
config.set('security', 'token_ttl', '3600')
config.set('security', 'ldap_server', 'localhost.localdomain')
config.set('security', 'ldap_port', '389')
config.set('security', 'user_base_dn', 'ou=people,ou=example,o=test')

config.add_section('db')
config.set('db', 'uri', 'sqlite://')
config.set('db', 'echo_queries', 'false')
config.set('db', 'pool_size', '50')

config.add_section('flask')
config.set('flask', 'propagate_exceptions', 'true')
config.set('flask', 'debug', 'false')
config.set('flask', 'strict_slashes', 'false')

config.add_section('logging')
config.set('logging', 'level', 'info')
config.set('logging', 'format', '%(asctime)s [%(name)s] %(levelname)s %('
                                'module)s:%(funcName)s:%(lineno)d - %('
                                'message)s')
config.set('logging', 'directory', defaults['ORLO_LOGDIR'])  # "disabled" for no
                                                             # log files

config.read(defaults['ORLO_CONFIG'])
